docs/source/daml/intro/9_Functional101.rst

```diff
@@ -0,0 +1,498 @@
+
+In the above, ``boom`` is an error.
+
+But while being proper control flow, ``case`` and ``if..else`` statements are also expressions in the sense that they result in a value when evaluated. You can actually see that in the function definitions above. Since each of the functions is defined just as a ``case`` or ``if`` statement, the value of the evaluated function is just the value of the ``case``/``if`` statement. Things that have a value have a type. The ``if..else`` expression in ``boolToInt2`` has type ``Int`` as that's what the function returns, the ``case`` expression in ``doError`` has type ``Bool``. To be able to give such expressions an unambiguous type, each branch needs to have the same type. The below function does not compile as one branch tries to return an ``Int`` and the other a ``Text``:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TYPE_ERROR_BEGIN
+  :end-before: -- TYPE_ERROR_END
+
+If we need functions that can return two (or more) types of things we need to encode that in the return type. For two possibilities, it's common to use the ``Either`` type:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- INT_OR_TEXT_BEGIN
+  :end-before: -- INT_OR_TEXT_END
+
+Branching in Actions
+~~~~~~~~~~~~~~~~~~~~
+
+The most common case where this becomes important is inside ``do`` blocks. Say we want to create a contract of one type in one case, and of another type in another case. Let's say we have two template types and want to write a function that creates an ``S`` if a condition is met, and a ``T`` otherwise.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- S_T_BEGIN
+  :end-before: -- S_T_END
+
+It would be tempting to write a simple ``if..else``, but it won't typecheck:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TYPE_ERROR2_BEGIN
+  :end-before: -- TYPE_ERROR2_END
+
+We have two options:
+
+1. Use the ``Either`` trick from above.
+2. Get rid of the return types.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- S_OR_T_BEGIN
+  :end-before: -- S_OR_T_END
+
+The latter is so common that there is a utility function in ``DA.Action`` to get rid of the return type: ``void : Functor f => f a -> f ()``.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- VOID_BEGIN
+  :end-before: -- VOID_END
+
+``void`` also helps express control flow of the type "Create a ``T`` only if a condition is met.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- CUSTOM_WHEN_BEGIN
+  :end-before: -- CUSTOM_WHEN_END
+
+Note that we still need the ``else`` clause of the same type ``()``. This pattern is so common, it's encapsulated in the standard library function ``DA.Action.when : (Applicative f) => Bool -> f () -> f ()``. 
```

Comment:
Maybe worth mentioning that `when` is also control flow in the sense that it short circuits.

Reply:
It only short circuits Action execution, though, not evaluation.

Reply:
it does short circuit evaluation. It’s hacked up in the compiler

Reply:
I had no idea!